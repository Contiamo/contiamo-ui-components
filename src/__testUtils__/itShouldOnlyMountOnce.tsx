import { mount } from "enzyme"
import * as React from "react"

/**
 * This calls the passed function on every `componentDidMount`
 * This is how we track is the parent un-mounts.
 * We can use hooks here because were cool.
 *
 */

function Child({ mockFN }: { mockFN: jest.Mock }) {
  React.useEffect(() => {
    mockFN()
  }, [])

  return <div />
}

/**
 * Must be a class in order to update state with enzyme.
 * We use this to re-render its child on every state change
 */
class TestWrapper extends React.Component<{ comp: React.FunctionComponent<{ count: number }> }, { count: number }> {
  public state = {
    count: 1,
  }

  public render() {
    return this.props.comp({ count: this.state.count })
  }
}

/**
 * A test hack to checks if a component does not re-mount on every render
 * Need this since we dont have access to life-cycle methods of the component we want to check,
 * we pass a child with a tracker because if the parent node un-mounts, all its children un-mount.
 *
 * Therefore, **we can only use this test for components that accept any children**.
 *
 *  ![TestHack](https://media1.tenor.com/images/7da78fe87cf5a457fbec3406e82d6cb3/tenor.gif)
 */
export const itShouldOnlyMountOnce = (componentName: string, Component: React.FunctionComponent<any>) => {
  const mockFn = jest.fn()

  const Inner = ({ count }: { count: number }) => {
    return (
      <React.Fragment>
        <button>Count:{count}</button>
        <Component>
          <Child mockFN={mockFn} />
        </Component>
      </React.Fragment>
    )
  }

  it(`[${componentName}] should only mount once`, () => {
    const m = mount(<TestWrapper comp={Inner} />)
    /// Lets Update the state 4 Times
    m.setState({ count: 2 })
    m.setState({ count: 3 })
    m.setState({ count: 4 })
    m.setState({ count: 5 })

    /// Component should have only mounted once
    expect(mockFn).toBeCalledTimes(1)
  })
}
