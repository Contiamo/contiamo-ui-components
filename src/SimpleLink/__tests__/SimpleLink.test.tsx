import * as React from "react"

import { render } from "enzyme"
import { itShouldOnlyMountOnce } from "../../__testUtils__/itShouldOnlyMountOnce"
import { SimpleLink as ThemelessSimpleLink } from "../../index"
import wrapDefaultTheme from "../../utils/wrap-default-theme"

const SimpleLink = wrapDefaultTheme(ThemelessSimpleLink)

describe("SimpleLink Component", () => {
  it("Should initialize properly", () => {
    expect(
      render(
        <SimpleLink left icon="Open" color="#000000">
          Custom color
        </SimpleLink>,
      ),
    ).toMatchSnapshot()
  })

  itShouldOnlyMountOnce("SimpleLink", SimpleLink)
})
