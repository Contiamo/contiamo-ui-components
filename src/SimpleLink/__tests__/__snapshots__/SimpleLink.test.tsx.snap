// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleLink Component Should initialize properly 1`] = `
<div
  class="css-ftf61z"
>
  <div
    class="css-p29iif"
  />
  <button
    aria-label="Custom color"
    class="css-5eru3u"
    role="button"
  >
    Custom color
    <svg
      class="css-1brfyce"
      fill="currentColor"
      height="12"
      viewBox="0 0 360 360"
      width="12"
    >
      <path
        d="M300,320.187l-260,0l0,-260l160,0l0,40l-120,0l0,180l180,0l0,-120l40,0l0,160Zm40,-200l-40,0l0,-31.715l-85.858,85.858l-28.284,-28.285l85.858,-85.858l-31.716,0l0,-40l100,0l0,100Z"
      />
    </svg>
  </button>
</div>
`;
